library/std/src/collections/hash/set.rs

```diff
@@ -998,6 +996,43 @@ where
     }
 }
 
+#[stable(feature = "std_collections_from_array", since = "1.53.0")]
+// Note: as what is currently the most convenient built-in way to construct
+// a HashSet, a simple usage of this function must not *require* the user
+// to provide a type annotation in order to infer the third type parameter
+// (the hasher parameter, conventionally "S").
+// To that end, this impl is defined using RandomState as the concrete
+// type of S, rather than being generic over `S: BuildHasher + Default`.
+// It is expected that users who want to specify a hasher will manually use
+// `with_capacity_and_hasher`.
+// If type parameter defaults worked on impls, and if type parameter
+// defaults could be mixed with const generics, then perhaps
+// this could be generalized.
+// See also the equivalent impl on HashMap.
+impl<T, const N: usize> From<[T; N]> for HashSet<T, RandomState>
+where
+    T: Eq + Hash,
+{
+    /// ```
+    /// use std::collections::HashSet;
+    ///
+    /// let set1 = HashSet::from([1, 2, 3, 4]);
+    /// let set2: HashSet<_> = [1, 2, 3, 4].into();
+    /// assert_eq!(set1, set2);
+    ///
+    /// let unordered_duplicates = HashSet::from([4, 1, 4, 3, 2]);
+    /// assert_eq!(set1, unordered_duplicates);
+    ///
+    /// // This next line must infer the hasher type parameter.
+    /// // If you make a change that causes this line to no longer infer,
+    /// // that's a problem!
+    /// let must_not_require_type_annotation = HashSet::from([1, 2]);
```

Comment:
Same goes for this, users does not need to know this detail.

Reply:
Agreed, this shouldn't be in the documentation.